File: android/app/src/main/java/com/example/earthwallet/chain/Dex.kt
package network.erth.wallet.chain

import com.google.protobuf.Any as ProtoAny
import cosmos.base.v1beta1.CoinOuterClass
import network.erth.earth.proto.dex.MsgAddLiquidity
import network.erth.earth.proto.dex.MsgRemoveLiquidity
import network.erth.earth.proto.dex.MsgSwap
import org.json.JSONObject

/**
 * x/dex (spoke-and-wheel AMM hubbed on ERTH) queries + messages.
 */
object Dex {

    data class Pool(
        val id: Long,
        val erthReserve: String, // uerth
        val tokenDenom: String,
        val tokenReserve: String,
        /**
         * 14-day-weighted swap volume, in real uerth.
         *
         * The chain does the weighting and hands over a plain number: a trade a
         * week ago counts (13/14)^7 of one made today. Nothing here has to age
         * it, and nothing here should try — an earlier version decayed this
         * client-side against a mechanism the chain had already replaced, and
         * the resulting APR drifted further out every day.
         *
         * At a steady trading rate it settles at about fourteen times the daily
         * volume, which is how an APR estimate works back to a daily figure.
         */
        val volumeErth: String = "0",
        /** Day index this pool last traded (block time / 86400). */
        val lastTradedDay: Long = 0,
    )

    /** Days in the rolling volume window. Mirrors types.VolumeDecayWindowDays. */
    const val VOLUME_WINDOW_DAYS = 14

    /** The LP share denom for a pool. Mirrors types.LPShareDenom. */
    fun shareDenom(poolId: Long): String = "dexlp/$poolId"

    /** How long withdrawn shares are escrowed before they pay out. */
    fun lpUnbondingSeconds(): Long {
        val (code, body) = EarthRest.get("/earth/dex/v1/params")
        if (code !in 200..299) return 0
        return JSONObject(body).getJSONObject("params")
            .optString("lp_unbonding_seconds", "0").toLongOrNull() ?: 0
    }

    private fun coin(denom: String, amount: String) =
        CoinOuterClass.Coin.newBuilder().setDenom(denom).setAmount(amount).build()

    // --- queries ---

    fun pools(): List<Pool> {
        val (code, body) = EarthRest.get("/earth/dex/v1/pool")
        if (code !in 200..299) return emptyList()
        val arr = JSONObject(body).optJSONArray("pool") ?: return emptyList()
        val out = ArrayList<Pool>(arr.length())
        for (i in 0 until arr.length()) {
            val p = arr.getJSONObject(i)
            out.add(
                Pool(
                    id = p.optString("pool_id", "0").toLong(),
                    erthReserve = p.getJSONObject("reserve_erth").getString("amount"),
                    tokenDenom = p.getJSONObject("reserve_token").getString("denom"),
                    tokenReserve = p.getJSONObject("reserve_token").getString("amount"),
                    volumeErth = p.optString("volume_erth", "0"),
                    lastTradedDay = p.optString("last_traded_day", "0").toLongOrNull() ?: 0L,
                )
            )
        }
        return out
    }

    /** A withdrawal waiting out its escrow. */
    data class Unbonding(
        val poolId: Long,
        val shares: String,
        /** Unix seconds at which it pays out on its own. */
        val completionTime: Long,
    )

    /**
     * Withdrawals this address has waiting.
     *
     * Between submitting one and it landing there is nothing in the balance to
     * show for it — the shares have left and the assets have not arrived — so
     * without this the week looks like the funds went nowhere.
     */
    fun unbondings(address: String): List<Unbonding> {
        val (code, body) = EarthRest.get("/earth/dex/v1/unbondings/$address")
        if (code !in 200..299) return emptyList()
        val arr = JSONObject(body).optJSONArray("unbondings") ?: return emptyList()
        return (0 until arr.length()).map { i ->
            val u = arr.getJSONObject(i)
            Unbonding(
                poolId = u.optString("pool_id", "0").toLongOrNull() ?: 0L,
                shares = u.optJSONObject("shares")?.optString("amount", "0") ?: "0",
                completionTime = u.optString("completion_time", "0").toLongOrNull() ?: 0L,
            )
        }
    }

    /** Pool that pairs ERTH with the given spoke token denom, if any. */
    fun poolForToken(tokenDenom: String): Pool? = pools().find { it.tokenDenom == tokenDenom }

    /** Swap fee as a percent string (e.g. "0.3"). */
    fun swapFeePercent(): String {
        val (code, body) = EarthRest.get("/earth/dex/v1/params")
        if (code !in 200..299) return "0"
        return JSONObject(body).getJSONObject("params").optString("swap_fee", "0")
    }

    // --- messages ---

    fun msgSwap(creator: String, tokenInDenom: String, tokenInAmount: String, denomOut: String, minOut: String): ProtoAny {
        val msg = MsgSwap.newBuilder()
            .setCreator(creator)
            .setTokenIn(coin(tokenInDenom, tokenInAmount))
            .setDenomOut(denomOut)
            .setMinAmountOut(minOut)
            .build()
        return EarthTx.anyOf("/earth.dex.v1.MsgSwap", msg)
    }

    fun msgAddLiquidity(creator: String, poolId: Long, denomA: String, amtA: String, denomB: String, amtB: String): ProtoAny {
        val msg = MsgAddLiquidity.newBuilder()
            .setCreator(creator).setPoolId(poolId)
            .setAmountA(coin(denomA, amtA)).setAmountB(coin(denomB, amtB))
            .build()
        return EarthTx.anyOf("/earth.dex.v1.MsgAddLiquidity", msg)
    }

    fun msgRemoveLiquidity(creator: String, poolId: Long, sharesDenom: String, sharesAmount: String): ProtoAny {
        val msg = MsgRemoveLiquidity.newBuilder()
            .setCreator(creator).setPoolId(poolId)
            .setShares(coin(sharesDenom, sharesAmount))
            .build()
        return EarthTx.anyOf("/earth.dex.v1.MsgRemoveLiquidity", msg)
    }
}

File: ios/EarthCore/Sources/EarthCore/Chain/Dex.swift
import Foundation

/// x/dex — the hub-and-spoke AMM. Every pool pairs ERTH with one token.
public enum Dex {

    /// Days in the rolling volume window. Mirrors `types.VolumeWindowDays`.
    public static let volumeWindowDays = 7

    /// The LP share denom for a pool. Mirrors `types.LPShareDenom`.
    public static func shareDenom(poolID: UInt64) -> String { "dexlp/\(poolID)" }

    public struct Pool: Sendable, Equatable, Identifiable {
        public let id: UInt64
        /// uerth.
        public let erthReserve: String
        public let tokenDenom: String
        public let tokenReserve: String
        /// ERTH-denominated swap volume, decayed daily over a rolling window.
        ///
        /// Not a plain total: each elapsed day multiplies it by 6/7, so at a
        /// steady rate it settles at roughly seven times the daily volume. It
        /// weights this pool's share of the LP reward stream, and an APR
        /// estimate has to work back from it.
        public let volume: String
        /// Day index the volume was last decayed to (block time / 86400).
        public let lastVolumeDay: Int64

        public init(
            id: UInt64,
            erthReserve: String,
            tokenDenom: String,
            tokenReserve: String,
            volume: String = "0",
            lastVolumeDay: Int64 = 0
        ) {
            self.id = id
            self.erthReserve = erthReserve
            self.tokenDenom = tokenDenom
            self.tokenReserve = tokenReserve
            self.volume = volume
            self.lastVolumeDay = lastVolumeDay
        }
    }

    /// A withdrawal waiting out its escrow.
    public struct Unbonding: Sendable, Equatable {
        public let poolID: UInt64
        public let shares: String
        /// Unix seconds at which it pays out on its own.
        public let completionTime: Int64
    }
}

public extension EarthClient {

    func pools() async -> [Dex.Pool] {
        guard let json = try? await rest.get("/earth/dex/v1/pool") else { return [] }
        return json.pool.array.compactMap { p in
            Dex.Pool(
                id: p.pool_id.uint64(default: 0),
                erthReserve: p.reserve_erth.amount.string(default: "0"),
                tokenDenom: p.reserve_token.denom.string(default: ""),
                tokenReserve: p.reserve_token.amount.string(default: "0"),
                volume: p.volume.string(default: "0"),
                lastVolumeDay: p.last_volume_day.int64(default: 0)
            )
        }
    }

    /// The pool pairing ERTH with the given spoke token, if there is one.
    func pool(forToken denom: String) async -> Dex.Pool? {
        await pools().first { $0.tokenDenom == denom }
    }

    /// Swap fee as a percent string, e.g. "0.3".
    func swapFeePercent() async -> String {
        guard let json = try? await rest.get("/earth/dex/v1/params") else { return "0" }
        return json.params.swap_fee.string(default: "0")
    }

    /// How long withdrawn shares are escrowed before they pay out.
    func lpUnbondingSeconds() async -> Int64 {
        guard let json = try? await rest.get("/earth/dex/v1/params") else { return 0 }
        return json.params.lp_unbonding_seconds.int64(default: 0)
    }

    /// Withdrawals this address has waiting.
    ///
    /// Between submitting one and it landing there is nothing in the balance to
    /// show for it — the shares have left and the assets have not arrived — so
    /// without this the week looks like the funds went nowhere.
    func unbondings(_ address: String) async -> [Dex.Unbonding] {
        guard let json = try? await rest.get("/earth/dex/v1/unbondings/\(address)")
        else { return [] }
        return json.unbondings.array.map { u in
            Dex.Unbonding(
                poolID: u.pool_id.uint64(default: 0),
                shares: u.shares.amount.string(default: "0"),
                completionTime: u.completion_time.int64(default: 0)
            )
        }
    }

    // --- messages ---

    func msgSwap(
        creator: String,
        tokenInDenom: String,
        tokenInAmount: String,
        denomOut: String,
        minAmountOut: String
    ) -> ProtoAny {
        Msg.Swap(
            creator: creator,
            tokenIn: Coin(denom: tokenInDenom, amount: tokenInAmount),
            denomOut: denomOut,
            minAmountOut: minAmountOut
        ).asAny(typeURL: Msg.Swap.typeURL)
    }

    func msgAddLiquidity(
        creator: String,
        poolID: UInt64,
        denomA: String, amountA: String,
        denomB: String, amountB: String
    ) -> ProtoAny {
        Msg.AddLiquidity(
            creator: creator,
            poolID: poolID,
            amountA: Coin(denom: denomA, amount: amountA),
            amountB: Coin(denom: denomB, amount: amountB)
        ).asAny(typeURL: Msg.AddLiquidity.typeURL)
    }

    func msgRemoveLiquidity(
        creator: String,
        poolID: UInt64,
        sharesDenom: String,
        sharesAmount: String
    ) -> ProtoAny {
        Msg.RemoveLiquidity(
            creator: creator,
            poolID: poolID,
            shares: Coin(denom: sharesDenom, amount: sharesAmount)
        ).asAny(typeURL: Msg.RemoveLiquidity.typeURL)
    }
}
